README
    eumm-upgrade is a tool to allow using new features of
    ExtUtils::MakeMaker without losing compatibility with older versions. It
    adds compatibility code to Makefile.PL and tries to automatically detect
    some properties like license, minimum Perl version required and
    repository used.

    Just run eumm-upgrade.pl in directory with Makefile.PL. Old file will be
    copied to Makefile.PL.bak. If you use Github, Internet connection is
    required.

    You need to check resulting Makefile.PL manually as transformation is done
    with regular expressions.

    If you need to declare number of spaces in indent in Makefile.PL, use
    following string at start of it (set 'c-basic-offset' to your value):

    # -*- mode: perl; c-basic-offset: 4; indent-tabs-mode: nil; -*-

    (c) Alexandr Ciornii



Changes
Revision history for App-EUMM-Upgrade

0.25
        - Option to skip searching for github parent repository
        - Use meta-spec 2 in case of github and bitbucket
        - When loading file from VERSION_FROM, also check if pod file is available
        - Detect WriteMakefile(%hash) and add new fields to %hash instead
        - Detect Dist::Zilla compatibility code and do not add own
        - Other small changes

0.24    2015-04-12
        - Perl::Meta is used
        - Better reindentation when outer block is removed
        - Better detection of exisiting BUILD_REQUIRES
        - Better detection of version-checking constructs
        - Better search for 'Perl' license
        - Scan 'author' section for license
        - Support for arrayref in 'AUTHOR'
        - Check for "package MY"
        - Support TEST_REQUIRES
        - Don't write compatibility code for perl < 5.005

0.22    2010-01-11
        - Better detection of version-checking constructs
        - Better search for 'Perl' license
        - Better detection of indent type
        - Other minor changes
        - Docs about new EUMM features

0.21    2009-12-18
        - First version released to CPAN


Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'App::EUMM::Upgrade',
    license             => 'gpl3',
    dist_author         => q{Alexandr Ciornii <alexchorny@gmail.com>},
    dist_version_from   => 'lib/App/EUMM/Upgrade.pm',
    requires => {
        'File::Slurp' => 0,
        'Text::FindIndent' => 0.08,
        'Github::Fork::Parent' => 0.20,
        'ExtUtils::MakeMaker' => 6.56,
        'Perl::Meta' => 0.01,
        'perl' => 5.006,
    },
    build_requires => {
        'Test::More' => 0.88,
    },
    script_files => [
        'eumm-upgrade.pl',
    ],
    meta_merge => {
        resources => {
            repository => 'http://github.com/chorny/EUMM-upgrade',
        },
        keywords => [
            'development','upgrade','EU::MM','Makefile.PL',
        ],
    },
    add_to_cleanup      => [ 'App-EUMM-Upgrade-*' ],
    create_makefile_pl => 'traditional',
    auto_configure_requires => 0,
);

$builder->create_build_script();


eumm-upgrade.pl
#!/usr/bin/perl

use 5.006;
use strict;
use warnings;

#License: GPL (may change in the future)

#use Perl6::Say;
use File::Slurp;
#require Module::Install::Repository;
#require Module::Install::Metadata;
use Text::FindIndent 0.08;
use Perl::Meta;
use App::EUMM::Upgrade;
use Getopt::Long;

my $noparent = 0;
GetOptions("noparent" => \$noparent) or die("Error in command line arguments\n");
print "will not search for github parent repository\n" if $noparent;

my $content=read_file('Makefile.PL') or die "Cannot find 'Makefile.PL'";
if ($content =~ /use inc::Module::Install/) {
  die "Module::Install is used, no need to upgrade";
}
if ($content =~ /WriteMakefile1\s*\(/) {
  print "Upgrade is already applied\n";
  exit;
}
if ($content !~ /\b(?:use|require) ExtUtils::MakeMaker/ or $content !~ /WriteMakefile\s*\(/) {
  die "ExtUtils::MakeMaker is not used";
}

sub process_file {
  my $content=shift;
  my $indentation_type = Text::FindIndent->parse($content,first_level_indent_only=>1);
  my $space_to_use;
  my $indent_str;
  if ($indentation_type =~ /^[sm](\d+)/) {
    print "Indentation with $1 spaces\n";
    $space_to_use=$1;
    $indent_str=' 'x$space_to_use;
  } elsif ($indentation_type =~ /^t(\d+)/) {
    print "Indentation with tabs, a tab should indent by $1 characters\n";
    $space_to_use=0;
    $indent_str="\t";
  } else {
    print "Indentation unknown, will use 4 spaces\n";
    $space_to_use=4;
    $indent_str=' 'x4;
  }

  my $compat_layer=<<'EOT';
sub WriteMakefile1 {  #Compatibility code for old versions of EU::MM. Written by Alexandr Ciornii, version 2. Added by eumm-upgrade.
    my %params=@_;
    my $eumm_version=$ExtUtils::MakeMaker::VERSION;
    $eumm_version=eval $eumm_version;
    die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
    die "License not specified" if not exists $params{LICENSE};
    if ($params{AUTHOR} and ref($params{AUTHOR}) eq 'ARRAY' and $eumm_version < 6.5705) {
        $params{META_ADD}->{author}=$params{AUTHOR};
        $params{AUTHOR}=join(', ',@{$params{AUTHOR}});
    }
    if ($params{TEST_REQUIRES} and $eumm_version < 6.64) {
        $params{BUILD_REQUIRES}={ %{$params{BUILD_REQUIRES} || {}} , %{$params{TEST_REQUIRES}} };
        delete $params{TEST_REQUIRES};
    }
    if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
        #EUMM 6.5502 has problems with BUILD_REQUIRES
        $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
        delete $params{BUILD_REQUIRES};
    }
    delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
    delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
    delete $params{META_MERGE} if $eumm_version < 6.46;
    delete $params{META_ADD} if $eumm_version < 6.46;
    delete $params{LICENSE} if $eumm_version < 6.31;

    WriteMakefile(%params);
}
EOT
  my $space=' 'x4;
  $content=~s/(WriteMakefile\()(?!\%)(\S)/$1\n$indent_str$2/;
  $content=remove_conditional_code($content,$indent_str);
  my @param;

  my $meta_modify_persent = 0;
  my $meta_modify_ver = 0;
  if ($content =~ m#META_MERGE['" ]\s*=>#) {
    $meta_modify_persent = 1;
    if ($content =~ m#['"]meta-spec['"]\s*=>\s*{\s*version\s*=>\s*2\s*}#s) {
      $meta_modify_ver = 2;
    }
  }

  my @resourses;
  my ($type, $repo) = find_repo();
  my $repo_string;
  if ($repo and $repo=~m#://#) {
    $repo = convert_url_to_public($repo);
    print "Repository found: $repo. Check that this is public URL.\n";
    unless ($noparent) {
      eval {
        require Github::Fork::Parent;
        $repo=Github::Fork::Parent::github_parent($repo);
      };
    }
    $repo_string = "repository => '$repo',";
  } else {
    $repo_string = "#repository => 'URL to repository here',";
  }
  my $web_url = convert_url_to_web($repo);
  if ($meta_modify_ver == 2 || (!$meta_modify_persent && $web_url))  {
    $meta_modify_ver = 2;
    push @resourses, "${space}${space}${space}repository => {";
    if ($repo) {
      push @resourses, "${space}${space}${space}${space}type => '$type',";
      push @resourses, "${space}${space}${space}${space}url => '$repo',";
      push @resourses, "${space}${space}${space}${space}web => '$web_url',";
    } else {
      push @resourses, "${space}${space}${space}${space}#see CPAN::Meta::Spec";
    }
    push @resourses, "${space}${space}${space}},";
  } else {
    push @resourses, "${space}${space}${space}$repo_string";
  }

  if ($content=~/\bVERSION_FROM['"]?\s*=>\s*['"]([^'"\n]+)['"]/ || $content=~/\bVERSION_FROM\s*=>\s*q\[?([^\]\n]+)\]/) {
    my $main_file=$1;
    my $main_file_pod = $1;
    my $main_file_content = eval { read_file($main_file) };
    if (!$main_file_content) {
      print "Cannot open $main_file\n";
    } else {
      if ($main_file_pod =~ s/\.pm$/\.pod/ && -e $main_file_pod) {
        $main_file_content .= "\n\n".eval { read_file($main_file_pod) };
      } else {
        $main_file_pod = '';
      }
      my @links=Perl::Meta::extract_bugtracker($main_file_content);
      if (@links==2) {
        my $remove;
        my $dist;
        foreach my $i (0.. $#links) {
          my $dist1;
          if ($links[$i] =~ m#^\Qhttp://rt.cpan.org/NoAuth/ReportBug.html?Queue=\E(\w+)#) {
            $remove = $i;
            $dist1 = $1;
          } elsif ($links[$i] =~ m#^\Qhttp://rt.cpan.org/NoAuth/Bugs.html?Dist=\E(\w+)#) {
            $dist1 = $1;
          } else {
            die "unknown bugtraker type $links[$i]";
          }
          if ($dist and $dist1 ne $dist) {
            die;
          }
        }
        splice(@links,$remove,1) if defined $remove;
      }
      if (@links==1) {
        my $bt=$links[0];
        print "Bugtracker found: $bt\n";
        if ($meta_modify_ver == 2)  {
          push @resourses, "${space}${space}${space}bugtracker => { 'web' => '$bt' },";
        } else {
          push @resourses, "${space}${space}${space}bugtracker => '$bt',";
        }
      } elsif (@links>1) {
        print "Too many links to bugtrackers found in $main_file\n";
      }
      if ($content !~ /\bLICENSE\s*=>\s*['"]/ and $content !~ /['"]LICENSE['"]\s*=>\s*['"]/) {
        my $l=Perl::Meta::extract_license($main_file_content);
        if ($l) {
          push @param,"    LICENSE => '$l',\n";
        } else {
          print "license not found\n";
        }
      }
      if ($content !~ /\bMIN_PERL_VERSION['"]?\s*=>\s*['"]?\d/) {
        my $version=Perl::Meta::extract_perl_version($main_file_content) ||
          Perl::Meta::extract_perl_version($content);
        if ($version) {
          push @param,"    MIN_PERL_VERSION => '$version',\n";
        }
      }
    }
  } else {
    print "VERSION_FROM not found\n";
    if ($content !~ /\bMIN_PERL_VERSION\s*=>\s*['"\d]/) {
      my $version=Perl::Meta::extract_perl_version($content);
      if ($version) {
        push @param,"    MIN_PERL_VERSION => '$version',\n";
      }
    }
  }

  if (@resourses) { # and $content !~ /\bMETA_MERGE\s*=>\s*\{/
    my $res=join("\n",@resourses);
    my $metaspec = '';
    #'meta-spec' => { version => 2 },
    
    if ($meta_modify_ver == 2 && !$meta_modify_persent)  {
      $metaspec="\n        'meta-spec' => { version => 2 },";
    }
    push @param,<<EOT;
    META_MERGE => {$metaspec
        resources => {
$res
        },
    },
EOT
  }

  if ($content !~ /\bTEST_REQUIRES['"]?\s*=>\s*\{/) {
    push @param,"    #TEST_REQUIRES => {\n"."    #},\n";
  }
  
  my $param='';
  if (@param) {
    $param="\n".join('',@param);
    $param = App::EUMM::Upgrade::apply_indent($param,4,$space_to_use);
    $param=~s/\s+$/\n/s;
  }
  $content = App::EUMM::Upgrade::add_new_fields($content, $param);
  #my $text2replace = 'WriteMakefile\(';
  #$content=~s/WriteMakefile\s*\(/WriteMakefile1($param/s;

  #$content=~s/[\r\n]+$//s;
  $compat_layer="\n\n".App::EUMM::Upgrade::apply_indent($compat_layer,4,$space_to_use);
  $content=~s/(package\s+MY; | __DATA__ | $ )/$compat_layer$1/sx;
  # |
  #
  return $content;
}

rename('Makefile.PL','Makefile.PL.bak');
write_file('Makefile.PL',process_file($content));

=pod

eumm-upgrade is a tool to allow using new features of ExtUtils::MakeMaker without losing
compatibility with older versions. It adds compatibility code to Makefile.PL and
tries to automatically detect some properties like license, minimum Perl version required and
repository used.

Just run eumm-upgrade.pl in directory with Makefile.PL. Old file will be copied to Makefile.PL.bak.
If you use Github, Internet connection is required.

You need to check resulting Makefile.PL manually as transformation is done
with regular expressions.

If you need to declare number of spaces in indent in Makefile.PL, use following string at start of
it (set 'c-basic-offset' to your value):

# -*- mode: perl; c-basic-offset: 4; indent-tabs-mode: nil; -*-

(c) Alexandr Ciornii
=cut

1;


lib/App/EUMM/Upgrade.pm
package App::EUMM::Upgrade;

use strict;
use warnings;

=head1 NAME

App::EUMM::Upgrade - Perl tool to upgrade ExtUtils::MakeMaker-based Makefile.PL

=head1 VERSION

Version 0.25

=cut

our $VERSION = '0.25';


=head1 SYNOPSIS

eumm-upgrade is a tool to allow using new features of ExtUtils::MakeMaker without losing
compatibility with older versions. It adds compatibility code to Makefile.PL and
tries to automatically detect some properties like license, minimum Perl version required and
repository used.

Just run eumm-upgrade.pl in directory with Makefile.PL. Old file will be copied to Makefile.PL.bak.
If you use Github, Internet connection is required.

You need to check resulting Makefile.PL manually as transformation is done
with regular expressions.

If you need to declare number of spaces in indent in Makefile.PL, use following string at start of
it (set 'c-basic-offset' to your value):

# -*- mode: perl; c-basic-offset: 4; indent-tabs-mode: nil; -*-

=head1 new EUMM features

LICENSE - shows license on search.cpan.org

META_MERGE - add something (like repository URL or bugtracker UTL) to META.yml. Repository and
bugtracker URL are used on search.cpan.org.

MIN_PERL_VERSION - minimum version of Perl required for module work. Not used currently, but will
be in the future.

CONFIGURE_REQUIRES - modules that are used in Makefile.PL and should be installed before running it.

TEST_REQUIRES - modules that are used in tests, but are not required by module
itself. Useful for ppm/OS package generaton and metadata parsing tools.

BUILD_REQUIRES - same as TEST_REQUIRES, but for building

AUTHOR - can be arrayref to allow several authors

=head1 AUTHOR

Alexandr Ciornii, C<< <alexchorny at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-app-eumm-upgrade at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=App-EUMM-Upgrade>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::EUMM::Upgrade


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-EUMM-Upgrade>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/App-EUMM-Upgrade>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/App-EUMM-Upgrade>

=item * Search CPAN

L<http://search.cpan.org/dist/App-EUMM-Upgrade/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009-2017 Alexandr Ciornii.

GPL v3

=cut

use Exporter 'import';
our @EXPORT=qw/remove_conditional_code find_repo convert_url_to_public convert_url_to_web/;
sub _indent_space_number {
  my $str=shift;
  $str=~/^(\s+)/ or return 0;
  my $ind=$1; 
  $ind=~s/\t/        /gs;
  return length($ind);
}

sub _unindent_t {
#  my $replace
#  die unless
}
sub _unindent {
  my $space_string_to_set=shift;
  my $text=shift;
  #print "#'$space_string_to_set','$text'\n";
  my @lines=split /(?<=[\x0A\x0D])/s,$text;
  use List::Util qw/min/;
  my $minspace=min(map {_indent_space_number($_)} @lines);
  my $s1=_indent_space_number($space_string_to_set);
  #die "$s1 > $minspace" if $s1 > $minspace;
  return $text if $s1==$minspace;
  #if (grep { $_ !~ /^$space_string_to_set/ } @lines) {
    
  #}
  #my $space_str
  my $line;
  my $i=0;
  foreach my $l (@lines) {
    next unless $l;
    if ($i==0) {
      $l =~ s/^\s+//;
      $i++;
      next;
    }
    unless ($l=~s/^$space_string_to_set//) {
      die "Text (line '$l') does not start with removal line ($space_string_to_set)";
    }
    next unless $l;
    if ($l=~m/^(\s+)/) {
      my $space=$1;
      if (!defined $line) {
        $line=$space;
        next;
      } else {
        if ($space=~/^$line/) {
          next;
        } elsif ($line=~/^$space/) {
          $line=$space;
          if ($line eq '') {
            #warn("line set to '' on line '$l'");
          }
        } else {
          die "Cannot find common start, on line '$l'";
        }
      }
    } else {
      return $text;
    }
  }
  if (!$line and $i>1) {
    die "Cannot find common start";
  }
  $i=0;
  foreach my $l (@lines) {
    next unless $l;
    if ($i==0) {
      $l="$space_string_to_set$l";
      $i++;
      next;
    }
    unless ($l=~s/^$line//) {
      die "Text (line '$l') does not start with calculated removal line ($space_string_to_set)";
    }
    $l="$space_string_to_set$l";
  }
  return (join("",@lines)."");

  #foreach
  #$text=~s/^(\s+)(\S)/_unindent_t(qq{$1},qq{$space_string_to_set}).qq{$2}/egm;
  
  #my $style=shift;
}

sub remove_conditional_code {
  my $content=shift;
  my $space=shift;
  $content=~s/(WriteMakefile\()(?!\%)(\S)/$1\n$space$2/;

  $content=~s/
  \(\s*\$\]\s*>=\s*5\.005\s*\?\s*(?:\#\#\s*\QAdd these new keywords supported since 5.005\E\s*)?
  \s+\(\s*ABSTRACT(?:_FROM)?\s*=>\s*'([^'\n]+)',\s*(?:\#\s*\Qretrieve abstract from module\E\s*)?
  \s+AUTHOR\s*=>\s*'([^'\n]+)'
  \s*\)\s*\Q: ()\E\s*\),\s+
  /ABSTRACT_FROM => '$1',\n${space}AUTHOR => '$2',\n/sx;

  my $eumm_version_check=qr/\$ ExtUtils::MakeMaker::VERSION\s+
          (?:g[et]\s+' [\d\._]+ ' \s* | >=?\s*[\d\._]+\s+) |
          eval\s*{\s*ExtUtils::MakeMaker->VERSION\([\d\._]+\)\s*}\s*
          /xs;
  $content=~s/
          ^(\s*)\(\s* $eumm_version_check
          \?\s+\(\s* #[\n\r]
          ( [ \t]*[^()]+? ) #main text, should not contain ()
           \s*
          \)\s*\:\s*\(\)\s*\),
  /_unindent($1,$2)/msxge;

  $content=~s/
          \(\s*\$\]\s* \Q>=\E \s* 5[\d\._]+ \s* \?\s*\( \s*
          ( [^()]+? ) ,? \s*
          \)\s*\:\s*\(\)\s*\),
  /$1,/sxg;
#    ($] >= 5.005 ?
#       (AUTHOR         => 'Stephen Hardisty <moowahaha@hotmail.com>') : ()),
  return $content;
}

sub _do_replace {
  my $spaces=shift;
  my $i_from=shift;
  my $i_to=shift;
  my $len=length($spaces);
  my $l1=int($len/$i_from);
  if ($i_to==0) {
    return "\t"x$l1;
  } else {
    return " " x ($l1*$i_to);
  }
}

sub apply_indent {
  my $content=shift;
  my $i_from=shift || die;
  my $i_to=shift;
  $content=~s/^((?:[ ]{$i_from})+)/_do_replace($1,$i_from,$i_to)/emg;
  return $content;
}

sub add_new_fields {
  my $content = shift;
  my $new_fields = shift;
  my $text2replace = 'WriteMakefile\(';
  if ($content =~ /WriteMakefile\(\s*(\%\w+)\s*\);/) {
    my $var_params = $1;
    $text2replace = qr/$var_params\s*=\s*\(\s*$/m;
    $content =~ s/($text2replace)/$1$new_fields/ or die "Cannot find $var_params initialization in Makefile.PL";
    $content =~ s/WriteMakefile\s*\(/WriteMakefile(/s;
  } else {
    $content =~ s/WriteMakefile\s*\(/WriteMakefile($new_fields/s;
  }
  return $content;
}



#_find_repo copied from Module::Install::Repository;
#by Tatsuhiko Miyagawa
#See Module::Install::Repository for copyright

sub _execute {
    my ($command) = @_;
    local $ENV{LC_ALL} = "C";
    `$command`;
}

sub find_repo {
  return _find_repo(\&_execute);
}

sub _find_repo {
    my ($execute) = @_;

    if (-e ".git") {
        # TODO support remote besides 'origin'?
        my $git_url =  '';
        if ($execute->('git remote show -n origin') =~ /URL: (.*)$/m) {
            # XXX Make it public clone URL, but this only works with github
            $git_url = $1;
            $git_url =~ s![\w\-]+\@([^:]+):!git://$1/!;
            return ('git', $git_url);
        } elsif ($execute->('git svn info') =~ /URL: (.*)$/m) {
            $git_url = $1;
        }
        return '' if $git_url =~ /\A\w+\z/;# invalid github remote might come back with just the remote name
        return ('git', $git_url);
    } elsif (-e ".svn") {
        if ($execute->('svn info') =~ /URL: (.*)$/m) {
            return ('svn', $1);
        }
    } elsif (-e "_darcs") {
        # defaultrepo is better, but that is more likely to be ssh, not http
        if (my $query_repo = `darcs query repo`) {
            if ($query_repo =~ m!Default Remote: (http://.+)!) {
                return ('darcs', $1);
            }
        }

        open my $handle, '<', '_darcs/prefs/repos' or return;
        while (<$handle>) {
            chomp;
            return ('darcs', $_) if m!^http://!;
        }
    } elsif (-e ".hg") {
        if ($execute->('hg paths') =~ /default = (.*)$/m) {
            my $mercurial_url = $1;
            $mercurial_url =~ s!^ssh://hg\@(bitbucket\.org/)!https://$1/!;
            return ('hg', $mercurial_url);
        }
    } elsif ($ENV{HOME} && -e "$ENV{HOME}/.svk") {
        # Is there an explicit way to check if it's an svk checkout?
        my $svk_info = `svk info` or return;
        SVK_INFO: {
            if ($svk_info =~ /Mirrored From: (.*), Rev\./) {
                return ('svk', $1);
            }

            if ($svk_info =~ m!Merged From: (/mirror/.*), Rev\.!) {
                $svk_info = `svk info /$1` or return;
                redo SVK_INFO;
            }
        }

        return;
    }
}

sub convert_url_to_public {
  my $url = shift;
  $url =~ s#^(?:ssh://|git://)?git\@(github\.com|bitbucket\.org)[:/]#https://$1/# and return $url;
  #ssh://git@bitbucket.org/shlomif/fc-solve.git

  $url =~ s#^(?:ssh://)hg\@(bitbucket\.org)/#https://$1/# and return $url;
  #ssh://hg@bitbucket.org/shlomif/app-notifier

  $url =~ s#^https://(\w+)\@(bitbucket\.org)/#https://$2/# and return $url;
  #ssh://hg@bitbucket.org/shlomif/app-notifier

  return $url;
}    

sub convert_url_to_web {
  my $url = shift;
  return unless $url;
  $url =~ s#^(?:https?|git)://(github\.com|bitbucket\.org)/(.*)\.git#https://$1/$2# and return $url;
  $url =~ s#^https?://(bitbucket\.org)/(.*)#https://$1/$2# and return $url;
  return;
}

1; # End of App::EUMM::Upgrade


Makefile.PL
require 5.006;
use ExtUtils::MakeMaker;
WriteMakefile1(
  'NAME' => 'App::EUMM::Upgrade',
  'VERSION_FROM' => 'lib/App/EUMM/Upgrade.pm',
  'ABSTRACT_FROM' => 'lib/App/EUMM/Upgrade.pm',
  'AUTHOR' => q{Alexandr Ciornii <alexchorny@gmail.com>},
  'CONFIGURE_REQUIRES' => {
                   'ExtUtils::MakeMaker' => '6.56',
  },
  'PREREQ_PM' => {
                   'ExtUtils::MakeMaker' => '6.56',
                   'File::Slurp' => 0,
                   'Github::Fork::Parent' => '0.24',
                   'Perl::Meta' => '0.03',
                   'Text::FindIndent' => '0.08',
  },
  TEST_REQUIRES => {
                   'Test::More' => '0.88',
  },
  LICENSE => 'gpl_3',
  MIN_PERL_VERSION => '5.006',
  META_MERGE => {
    'meta-spec' => { version => 2 },
    resources => {
      repository => {
        type => 'git',
        url => 'git://github.com/chorny/EUMM-upgrade.git',
        web => 'https://github.com/chorny/EUMM-upgrade',
      },
    },
    keywords => [
      'development','upgrade','EU::MM','Makefile.PL',
    ],
  },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [
                   'eumm-upgrade.pl'
                 ],
  'PL_FILES' => {},
  $^O =~/win/i ? (
    dist => {
      TAR      => 'ptar',
      TARFLAGS => '-c -C -f',
    },
  ) : (),
)
;

sub WriteMakefile1 {  #Compatibility code for old versions of EU::MM. Written by Alexandr Ciornii, version 2. Added by eumm-upgrade.
  my %params=@_;
  my $eumm_version=$ExtUtils::MakeMaker::VERSION;
  $eumm_version=eval $eumm_version;
  die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
  die "License not specified" if not exists $params{LICENSE};
  if ($params{AUTHOR} and ref($params{AUTHOR}) eq 'ARRAY' and $eumm_version < 6.5705) {
    $params{META_ADD}->{author}=$params{AUTHOR};
    $params{AUTHOR}=join(', ',@{$params{AUTHOR}});
  }
  if ($params{TEST_REQUIRES} and $eumm_version < 6.64) {
    $params{BUILD_REQUIRES}={ %{$params{BUILD_REQUIRES} || {}} , %{$params{TEST_REQUIRES}} };
    delete $params{TEST_REQUIRES};
  }
  if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
    delete $params{BUILD_REQUIRES};
  }
  delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
  delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
  delete $params{META_MERGE} if $eumm_version < 6.46;
  delete $params{META_ADD} if $eumm_version < 6.46;
  delete $params{LICENSE} if $eumm_version < 6.31;

  WriteMakefile(%params);
}
